### ISSUE TITLE
Is Rails still supported?

### ISSUE THREAD

**Issue by patleb** [thoughtbot/bourbon#925]
I was trying the version 5 of Bourbon and assets requirements do not work anymore, I have to go through bower-rails to make it work. The suspected commit which remove Rails integration is this one 73b09f618bd3189bf5af6f639e3df07680ed4fc7.

**Comment by tysongach** [comment]
Hi @patleb!

Yes, we still support Rails (4.2+ to be exact, with Bourbon 5). What beta number of Bourbon 5 are you using?

We did have a bug in v5.0.0.beta.3. If you’re using v5.0.0.beta.5/v5.0.0.beta.6, you should be fine.

Re: commit 73b09f618bd3189bf5af6f639e3df07680ed4fc7: We removed an old Rails Engine which was needed back with Rails 3. That commit did introduce a `SASS_PATH` issue, though, which was reverted in a462ed2ac92db0e40ea2a7716eb7f5628e3ea523.

**Comment by patleb** [comment]
The complete stack is Rails 5, Sprockets 3 (with sassc-rails and other stuff to make it fast) and I was on the master branch (so v5.0.0.beta.6). I took a look and `SASS_PATH` isn't used in sassc-rails. Not sure though if it's a good idea to set directly an env variable instead of adding a path to `Sass.load_paths` or using some other configs in the app context as fallback (sass-rails does it). Engines/Railties are still pretty much in use for scoping/including gems assets, it's old only if you use something else :wink:

**Comment by tysongach** [comment]
Apologies, the Rails stuff is a bit out of my wheelhouse.

- What’s the exact error that you’re getting?
- What does your `application.scss` look like? You removed the Sprockets directives, correct?
- The the code open-source somewhere we could see it?

I’ll have to do some more digging…

**Comment by patleb** [comment]
No apologies needed, I understand that Rails Assets Pipeline and Engines aren't straightforward to grasp without having played (suffered) with before. I've made a commit for re-integrating bourbon within Rails assets pipeline if you wanna a take a look 1d49cfb3b379eeac723f62c29c056fe357686aa5.

**Comment by patleb** [comment]
Let me know if you want a pull request.

**Comment by tysongach** [comment]
@patleb Can you further explain what that does? I’m not sure I grasp…

**Comment by patleb** [comment]
Yep, sure. Sprockets has sass-rails (or sassc-rails, in my case) to define the Sass preprocessor and both use `config.assets.paths` to tell Sass where to find the assets (`require` and `@import` works similarly since sass-rails customizes the `@import` directive). So basically, my modification tells Rails to add the `core` directory to its configuration when the Bourbon::Engine is loaded.

**Comment by tysongach** [comment]
@patleb So, you’re using Sprockets and `require` directives in your `application.scss`? We’ve had issues in the past with Sprockets because it doesn’t handle things the same way Sass’s `@import` does. More info here: https://blog.pivotal.io/labs/labs/structure-your-sass-files-with-import

And [from the Rails guides](http://guides.rubyonrails.org/asset_pipeline.html):

> If you want to use multiple Sass files, you should generally use the Sass @import rule instead of these Sprockets directives. When using Sprockets directives, Sass files exist within their own scope, making variables or mixins only available within the document they were defined in.

I’m not sure where I see where sass-rails is _part of_ Sprockets? Also, how does sass-rails customize the `@import` directive?

For reference here’s a bit of history on the changes we spoke about in Bourbon:

- PR that moved away from `Sass.load_paths` to `SASS_PATH `: https://github.com/thoughtbot/bourbon/pull/431
- Sass 3.3 change log when `SASS_PATH` was introduced: https://github.com/sass/sass/blob/stable/doc-src/SASS_CHANGELOG.md#backwards-incompatibilities----must-read

Also worth noting that other libraries are using `SASS_PATH`, like Susy: https://github.com/oddbird/susy/blob/master/lib/susy.rb

**Comment by patleb** [comment]
Nop, I'm well aware of drawbacks with `require` in sass, I'm using `@import` and sass-rails is not part of Sprockets but uses it and registers a `SassImporter` as a new kind of preprocessor so we could use `.erb` without specifying the extension and allow a generic directory import like this `@import 'core/**/*'`. I understand how `Sass.load_paths` could be problematic, but it's independent of Rails Assets Pipeline which is implemented by sprockets-rails which itself uses heavily `config.assets.paths`, not `Sass.load_paths`. Here are some references:

- [issue with Susy and how to make it work](https://github.com/sass/sassc-rails/issues/48)

- [mention where @import is overriden](https://github.com/rails/sass-rails#important-note)

- [SassImporter registered](https://github.com/rails/sass-rails/blob/28907966729276f9585b79ca48c28569e37a1d6c/lib/sass/rails/railtie.rb#L57)

- [@import functionalities added](https://github.com/rails/sass-rails/blob/28907966729276f9585b79ca48c28569e37a1d6c/lib/sass/rails/importer.rb#L35)

**Comment by tysongach** [comment]
@patleb I guess I’m just trying to understand where discrepancy is. We use Bourbon in Rails apps everyday here at thoughtbot and haven’t had an issue. I even just spun up a quick Rails 5 app to test and all seemed fine. This was using sass-rails, not sassc-rails.

So this seems to be an issue with sassc-rails, specifically, right? Is there a reason they apparently don’t support `SASS_PATH`? The issue you linked to has been open for awhile with no follow-up from the contributor that commented, which leads me to think it’s something they plan to work on?

Also, just to confirm: you’re seeing an error related to not being able to load Bourbon?

**Comment by patleb** [comment]
Yes, exactly, it's a loading error, the problem seems to be on sassc-rails side (if `SASS_PATH` is supposed to be the standard way to configure load paths). I don't know why it doesn't work, I might take a deeper look eventually. Thx for your time.

**Comment by patleb** [comment]
I verified and there is no implementation with `SASS_PATH` in [sassc](https://github.com/sass/sassc), but there was a discussion about it [here](https://github.com/sass/libsass/issues/90). I doubt that it'll be added soon, so it comes down to the question : do you want bourbon to support sassc-rails or not?

**Comment by tysongach** [comment]
@patleb I don’t think it’s quite that simple. It seems like core Sass contributors are indicating that sassc _should_ be supporting `SASS_PATH`:

> …to add to sassc which would be the appropriate place to handle interactions with the operating system.

and

> Agree, this should be supported by the libsass frontends. We should have a guide of things that frontends are required to do in order to be a compliant frontend.

---

> do you want bourbon to support sassc-rails or not?

Of course we do, but introducing complexities here within our code base to account for sassc seemingly not following the guidance of the Sass team feels like the wrong path forward. It’s also worth noting that SassC::Rails is a wrapper around SassC-Ruby, not SassC (or so it appears).

I’m also a bit confused because I have successfully used sassc with Bourbon many times. In fact, here’s the source for the new WIP Bourbon site, which uses Middleman and sassc: https://github.com/thoughtbot/bourbon.io/blob/master/Gemfile#L16

Have you thought about opening an issue on SassC or SassC-Ruby?

**Comment by patleb** [comment]
You're right, it's SassC-Ruby, my mistake. I just opened an issue about it [here](https://github.com/sass/sassc-ruby/issues/46) and there is [my sample app](https://github.com/patleb/bourbon_sample_app) to test the scenario if you want (the example you showed me isn't a Rails app).

Even if this is considered as an added complexity (which I'm not convinced considering the amount of time I passed looking into how all the sass stuff gets loaded), bourbon import won't work until sassc is modified and the dependency will require a minimal version of sassc (which you might have to specify either in the readme or by adding some logic to restrict the version or... just let the compiler output an import error). If sassc-ruby becomes eventually aware of SASS_PATH, it might be necessary to adapt sassc-rails as well and I'm really not convinced using an env variable to modify the internal behavior of the sass compiler was the right call when other explicit options are available : side-effects of an env variable are way harder to debug then a config in a restricted context like the sass compilers have and it's an obscure API.

The way I see it is that conditionally registering this gem as an engine make it work seamlessly with any sass compilers in the context of Rails and Sprockets (which was my initial concern about Rails support). Also, I've seen countless time librairies conditionally registering as an engine when Rails support was advertise as such, I'm a bit confused myself to see it skipped here.
